File: js/group_anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
const groupAnagrams = (strs) => {
  const groups = new Map();
  const result = [];

  for (const word of strs) {
    const mapKey = word.split('').sort().join();

    if (groups.has(mapKey)) {
      const arr = groups.get(mapKey);
      arr.push(word);
    } else {
      groups.set(mapKey, [word]);
    }
  }

  for (const [_, value] of groups) {
    result.push(value);
  }

  // return result;
  return Array.from(groups.values())
};

// Dummy Data
const words = ["eat", "tea", "tan", "ate", "nat", "bat"];
console.log(groupAnagrams(words));

File: ruby/group_anagrams.rb
def group_anagrams(strs)
  groups = Hash.new { |hash, key| hash[key] = [] }

  strs.each do |word|
    key = word.chars.sort.join
    groups[key] << word
  end

  groups.values
end

def group_anagrams_oneline(strs)
  strs.group_by { |word| word.chars.sort.join }.values
end

words = ["eat", "tea", "tan", "ate", "nat", "bat"]
p group_anagrams(words)
p group_anagrams_oneline words
# Expected Output: [["eat", "tea", "ate"], ["tan", "nat"], ["bat"]]
